text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,137 @@
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+Fields may be declared `unsafe`. Unsafe fields may only be mutated (excluding interior mutability)
+or initialized in an unsafe context. Reading the value of an unsafe field may occur in either safe
+or unsafe contexts. An unsafe field may be relied upon as a safety invariant in other unsafe code.
+
+# Motivation
+
+Emphasis on safety is a key strength of Rust. A major point here is that any code path that can
+result in undefined behavior must be explicitly marked with the `unsafe` keyword. However, the
+current system is insufficient. While Rust provides the `unsafe` keyword at the function level,
+there is currently no mechanism to mark fields as `unsafe`.
+
+For a real-world example, consider the `Vec` type in the standard library. It has a `len` field that
+is used to store the number of elements present. Setting this field is exposed publicly in the
+`Vec::set_len` method, which has safety requirements:
+
+- `new_len` must be less than or equal to `capacity()`.
+- The elements at `old_len..new_len` must be initialized.
+
+This field is safe to read, but unsafe to mutate or initialize due to the invariants. These
+invariants cannot be expressed in the type system, so they must be enforced manually. Failure to do
+so may result in undefined behavior elsewhere in `Vec`.
+
+By introducing unsafe fields, Rust can improve the situation where a field that is otherwise safe is
+used as a safety invariant.
+
+# Guide-level explanation
+
+Fields may be declared `unsafe`. Unsafe fields may only be initialized or accessed mutably in an
+unsafe context. Reading the value of an unsafe field may occur in either safe or unsafe contexts. An
+unsafe field may be relied upon as a safety invariant in other unsafe code.
+
+Here is an example to illustrate usage:
+
+```rust
+struct Foo {
+    safe_field: u32,
+    /// Safety: Value must be an odd number.
+    unsafe unsafe_field: u32,
+}
+
+// Unsafe field initialization requires an `unsafe` block.
+// Safety: `unsafe_field` is odd.
+let mut foo = unsafe {
+    Foo {
+        safe_field: 0,
+        unsafe_field: 1,
+    }
+};
+
+// Safe field: no unsafe block.
+foo.safe_field = 1;
+
+// Unsafe field with mutation: unsafe block is required.
+// Safety: The value is odd.
+unsafe { foo.unsafe_field = 3; }
+
+// Unsafe field without mutation: no unsafe block.
+println!("{}", foo.unsafe_field);
+```
+
+For a full description of where a mutable access is considered to have occurred (and why), see
+[RFC 3323]. Keep in mind that due to reborrowing, a mutable access of an unsafe field is not
+necessarily explicit.
+
+[RFC 3323]: https://rust-lang.github.io/rfcs/3323-restrictions.html#where-does-a-mutation-occur
+
+```rust
+fn change_unsafe_field(foo: &mut Foo) {
+    // Safety: An odd number plus two remains an odd number.
+    unsafe { foo.unsafe_field += 2; }
```

Comment:
suggestion: include a section talking about how both of the following are reasonable types:

```rust
/// This type has a *correctness* invariant that it holds an even number,
/// but it's not something on which `unsafe` code is allowed to rely.
pub struct Even(u32);

impl Even {
    pub fn new(x: u32) -> Option<Self>;
    pub fn new_i_promise(x: u32) -> Self;
}
```

```rust
/// # Safety
///
/// The value of this type must always be an even number.
/// (`unsafe` code is allowed to rely on that fact.)
pub unsafe struct Even(u32);

impl Even {
    pub fn new(x: u32) -> Option<Self>;
    pub unsafe fn new_unchecked(x: u32) -> Self;
}
```

and it's up to the type author to decide which is appropriate for the expected uses of the type.